generics/setttl.go

```diff
@@ -0,0 +1,75 @@
+package generics
+
+import (
+	"cmp"
+	"sort"
+	"time"
+
+	"golang.org/x/exp/maps"
+)
+
+// SetWithTTL is a unique set of items with a TTL (time to live) for each item.
+// After the TTL expires, the item is automatically removed from the set.
+type SetWithTTL[T cmp.Ordered] struct {
+	Items map[T]time.Time
+	TTL   time.Duration
+}
+
+// NewSetWithTTL returns a new SetWithTTL with elements `es` and a TTL of `ttl`.
+func NewSetWithTTL[T cmp.Ordered](ttl time.Duration, es ...T) SetWithTTL[T] {
+	s := SetWithTTL[T]{
+		Items: make(map[T]time.Time, len(es)),
+		TTL:   ttl,
+	}
+	s.Add(es...)
+	return s
+}
+
+// Add adds elements `es` to the SetWithTTL.
+func (s SetWithTTL[T]) Add(es ...T) {
+	for _, e := range es {
+		s.Items[e] = time.Now().Add(s.TTL)
```

Comment:
If we still want to use `clockwork` package as a unified way to management time in Refinery, we will need to inject that into this generic implementation of SetWithTTL

Reply:
We can't really do injection here; this is not a Start-able component. I also didn't want to make it dependent on passing a Clock in whenever you need one. 

What do you think of giving it a Clock member that defaults to a Real clock, and you have to manually override it in tests?

Reply:
Yes, that's actually what I was thinking too